sdk/opendp/whitenoise/synthesizers/pytorch/nn/dpctgan.py

```diff
@@ -0,0 +1,331 @@
+                    cross_entropy = self._cond_loss(fake, c1, m1)
+
+                if self.loss=='cross_entropy':
+                    label_g = torch.full((int(self.batch_size/self.pack),), REAL_LABEL, device=self.device)
+                    #label_g = torch.full(int(self.batch_size/self.pack,),1,device=self.device)
+                    loss_g = criterion(y_fake, label_g)
+                    loss_g = loss_g + cross_entropy
+                else:
+                    loss_g = -torch.mean(y_fake) + cross_entropy
+
+                optimizerG.zero_grad()
+                loss_g.backward()
+                optimizerG.step()
+
+                if not self.disabled_dp:
+                    if self.loss == 'cross_entropy':
+                        autograd_grad_sample.clear_backprops(discriminator)
+                    else:
+                        for p in discriminator.parameters():
+                            if hasattr(p, "grad_sample"):
+                                del p.grad_sample 
+
+                    if self.target_delta is None:
+                        self.target_delta = 1/train_data.shape[0]
+
+                    epsilon, best_alpha = optimizerD.privacy_engine.get_privacy_spent(self.target_delta)
+                    
+                    self.epsilon_list.append(epsilon)
+                    self.alpha_list.append(best_alpha)
+                    #if self.verbose:
+                        
+            
+            if not self.disabled_dp:
+                if self.budget < epsilon:
+                    break
+            self.loss_d_list.append(loss_d)
+            self.loss_g_list.append(loss_g)
+            if self.verbose:
+                print("Epoch %d, Loss G: %.4f, Loss D: %.4f" %
+                  (i + 1, loss_g.detach().cpu(), loss_d.detach().cpu()),
+                  flush=True)
+                print ('epsilon is {e}, alpha is {a}'.format(e=epsilon, a = best_alpha))
+
+        return self.loss_d_list, self.loss_g_list, self.epsion_list, self.alpha_list
+
+
+    def generate(self, n):
+        self.generator.eval()
+
+        #output_info = self.transformer.output_info
+        steps = n // self.batch_size + 1
+        data = []
+        for i in range(steps):
+            mean = torch.zeros(self.batch_size, self.embedding_dim)
+            std = mean + 1
+            fakez = torch.normal(mean=mean, std=std).to(self.device)
+
+            condvec = self.cond_generator.sample_zero(self.batch_size)
+            if condvec is None:
+                pass
+            else:
+                c1 = condvec
+                c1 = torch.from_numpy(c1).to(self.device)
+                fakez = torch.cat([fakez, c1], dim=1)
+
+            fake = self.generator(fakez)
+            fakeact = self._apply_activate(fake)
+            data.append(fakeact.detach().cpu().numpy())
+
+        data = np.concatenate(data, axis=0)
+        data = data[:n]
+        return self.transformer.inverse_transform(data, None)
```

Comment:
return data. inverse_transform(data, None) should get handled by PytorchDPSynthesizer.generate(). The idea is to allow a user to choose any preprocessor and GAN they want without forcing them to use either.